src/main/java/com/hotels/bdp/circustrain/bigquery/partition/PartitionValueFormatter.java

```diff
@@ -15,36 +15,54 @@
  */
 package com.hotels.bdp.circustrain.bigquery.partition;
 
-import java.util.List;
-import java.util.Set;
+import com.google.cloud.bigquery.FieldValue;
 
-import org.apache.hadoop.hive.metastore.api.FieldSchema;
+class PartitionValueFormatter {
 
-import com.google.common.collect.ImmutableSet;
+  private final String partitionKeyType;
+  private final FieldValue partitionFieldValue;
 
-class PartitionValueFormatter {
+  PartitionValueFormatter(FieldValue partitionFieldValue, String partitionKeyType) {
+    this.partitionFieldValue = partitionFieldValue;
+    this.partitionKeyType = partitionKeyType.toLowerCase();
+  }
 
-  private static final Set<String> NEEDS_FORMATTING = ImmutableSet.of("string");
-  private boolean format = false;
+  String formatValue() {
+    switch (partitionKeyType) {
+    case "string":
+    case "date":
+      return getValueWithQuotes();
+    case "timestamp":
+      return getTimestampValue();
+    default:
+      return partitionFieldValue.getStringValue();
+    }
+  }
 
-  PartitionValueFormatter(String partitionKey, List<FieldSchema> cols) {
-    partitionKey = partitionKey.toLowerCase().trim();
-    for (FieldSchema col : cols) {
-      String name = col.getName().toLowerCase().trim();
-      String type = col.getType().toLowerCase().trim();
+  private String getValueWithQuotes() {
+    return String.format("\"%s\"", partitionFieldValue.getStringValue());
+  }
 
-      if (name.equals(partitionKey) && NEEDS_FORMATTING.contains(type)) {
-        format = true;
-        break;
-      }
+  private String getTimestampValue() {
+    if (isNumber()) {
+      return getTimestampFromNumber();
+    } else {
+      throw new IllegalStateException(
+          "Expected to get number from BigQuery for timestamp column but got " + partitionFieldValue);
     }
   }
 
-  String format(String partitionValue) {
-    partitionValue = partitionValue.trim();
-    if (!format) {
-      return partitionValue;
+  private String getTimestampFromNumber() {
+    return String.format("TIMESTAMP_MICROS(%s)", partitionFieldValue.getTimestampValue());
+  }
+
+  private boolean isNumber() {
+    try {
+      Double.valueOf(partitionFieldValue.getStringValue());
+      return true;
+    } catch (Exception e) {
```

Comment:
I'd prefer to just catch `NumberFormatException` here. I realise we could also get a NPE which this currently covers but would that ever happen?

Reply:
Technically, it can happen, because a FieldValue stores an Object as the value, but I'll test and see how a null value in the table is returned for FieldValue. I'll get back to this as soon as I fix a some things I just broke :)

Reply:
So it can be null when the partition filter is "column IS NULL", in which case, yes, a NullPointerException is returned.